include/hal_core/utilities/enums.h

```diff
@@ -10,58 +10,60 @@ namespace hal
      * Container storing the string description of an enum.
      * Must be declared for every enum to be supported.
      */
-    template<typename T>
+    template<typename T, typename = typename std::enable_if<std::is_enum<T>::value, T>::type>
     struct EnumStrings
     {
         static std::vector<std::string> data;
     };
 
     /**
      * Translates an enum value into its string representation.
-     * 
+     *
      * @param[in] e - The enum value.
      * @returns The string representation.
      */
-    template<typename T>
-    std::string enum_to_string(T const& e)
+    template<typename T, typename = typename std::enable_if<std::is_enum<T>::value, T>::type>
+    std::string enum_to_string(T e)
     {
-        return EnumStrings<T>::data.at(static_cast<int>(e));
+        auto index = static_cast<size_t>(e);
+        if (index >= EnumStrings<T>::data.size())
+        {
+            throw std::runtime_error("no string for enum with value '" + std::to_string(index) + "'");
+        }
+        return EnumStrings<T>::data.at(index);
     }
 
     /**
      * Translates a string into an enum value if possible.
      * Assumes the last enum entry to represent an invalid state, as it defaults to this value.
-     * 
+     *
      * @param[in] str - The string.
      * @returns The enum value.
      */
-    template<typename T>
+    template<typename T, typename = typename std::enable_if<std::is_enum<T>::value, T>::type>
     T enum_from_string(const std::string& str)
     {
-        u32 i;
-        for (i = 0; i < EnumStrings<T>::data.size()-1; i++)
+        for (size_t i = 0; i < EnumStrings<T>::data.size(); i++)
         {
             if (EnumStrings<T>::data.at(i) == str)
             {
-                break;
+                return static_cast<T>(i);
             }
         }
 
-        return static_cast<T>(i);
+        return static_cast<T>(EnumStrings<T>::data.size() - 1);
```

Comment:
imo, old style was difficult to read.
I strongly advise against defaulting to the last entry, if the string is not found, since the mechanism is available to all plugins etc.
While the behavior is documented, it's still unintuitive.
I would go for an exception or default-in-case-of-error value.